File: programs/ball-sort/src/types/events.rs
use anchor_lang::prelude::*;

#[event]
pub struct GameConfigInitialized {
    pub authority: Pubkey,
    pub treasury: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct GameConfigUpdated {
    pub authority: Pubkey,
    pub field: String,
    pub timestamp: i64,
}

#[event]
pub struct PermissionCreated {
    pub player: Pubkey,
    pub puzzle_board: Pubkey,
    pub puzzle_stats: Pubkey,
    pub puzzle_board_permission: Pubkey,
    pub puzzle_stats_permission: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct PuzzleInitialized {
    pub player: Pubkey,
    pub puzzle_board: Pubkey,
    pub puzzle_stats: Pubkey,
    pub num_tubes: u8,
    pub balls_per_tube: u8,
    pub difficulty: u8,
    pub timestamp: i64,
}

#[event]
pub struct PuzzleStarted {
    pub puzzle_board: Pubkey,
    pub player: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct PuzzleFinalized {
    pub player: Pubkey,
    pub puzzle_board: Pubkey,
    pub puzzle_stats: Pubkey,
    pub move_count: u32,
    pub undo_count: u32,
    pub difficulty: u8,
    pub timestamp: i64,
}

#[event]
pub struct PuzzleAbandoned {
    pub player: Pubkey,
    pub puzzle_board: Pubkey,
    pub puzzle_stats: Pubkey,
    pub move_count: u32,
    pub undo_count: u32,
    pub difficulty: u8,
    pub timestamp: i64,
}

#[event]
pub struct TournamentCreated {
    pub tournament: Pubkey,
    pub authority: Pubkey,
    pub entry_fee: u64,
    pub difficulty: u8,
    pub end_time: i64,
    pub treasury_fee_bps: u16,
    pub timestamp: i64,
}

#[event]
pub struct TournamentJoined {
    pub tournament: Pubkey,
    pub player: Pubkey,
    pub timestamp: i64,
}

#[event]
pub struct TournamentResultRecorded {
    pub tournament: Pubkey,
    pub player: Pubkey,
    pub weight: u128,
    pub elapsed_secs: u64,
    pub move_count: u32,
    pub timestamp: i64,
}

#[event]
pub struct TournamentClosed {
    pub tournament: Pubkey,
    pub total_entries: u32,
    pub total_completers: u32,
    pub prize_pool: u64,
    pub timestamp: i64,
}

#[event]
pub struct PrizeClaimed {
    pub tournament: Pubkey,
    pub player: Pubkey,
    pub amount: u64,
    pub timestamp: i64,
}

#[event]
pub struct PuzzleDelegated {
    pub puzzle_board: Pubkey,
    pub player: Pubkey,
    pub validator: Option<Pubkey>,
    pub timestamp: i64,
}

File: backend/src/types/events.ts
// ─── Helius Enhanced Transaction ─────────────────────────────────────────────

export interface HeliusAccountData {
  account: string;
  nativeBalanceChange: number;
  tokenBalanceChanges: unknown[];
}

export interface HeliusNativeTransfer {
  fromUserAccount: string;
  toUserAccount: string;
  amount: number;
}

export interface HeliusInstruction {
  accounts: string[];
  data: string;
  programId: string;
  innerInstructions: unknown[];
}

export interface HeliusEnhancedTx {
  signature: string;
  slot: number;
  timestamp: number;
  fee: number;
  feePayer: string;
  computeUnitsConsumed: number;
  transactionError: unknown | null;
  type: string;
  source: string;
  accountData: HeliusAccountData[];
  nativeTransfers: HeliusNativeTransfer[];
  tokenTransfers: unknown[];
  instructions: HeliusInstruction[];
  innerInstructions: unknown[];
  events: Record<string, unknown>;
}

// ─── Anchor Event Payloads (from IDL types section) ──────────────────────────

export interface PuzzleFinalizedEvent {
  player: string;
  puzzleBoard: string;
  puzzleStats: string;
  moveCount: number;
  undoCount: number;
  difficulty: number;
  timestamp: number;
}

export interface PuzzleAbandonedEvent {
  player: string;
  puzzleBoard: string;
  puzzleStats: string;
  moveCount: number;
  undoCount: number;
  difficulty: number;
  timestamp: number;
}

export interface PuzzleInitializedEvent {
  player: string;
  puzzleBoard: string;
  puzzleStats: string;
  numTubes: number;
  ballsPerTube: number;
  difficulty: number;
  timestamp: number;
}

export interface TournamentCreatedEvent {
  tournament: string;
  authority: string;
  entryFee: bigint;
  difficulty: number;
  endTime: number;
  treasuryFeeBps: number;
  timestamp: number;
}

export interface TournamentJoinedEvent {
  tournament: string;
  player: string;
  timestamp: number;
}

export interface TournamentClosedEvent {
  tournament: string;
  totalEntries: number;
  totalCompleters: number;
  prizePool: bigint;
  timestamp: number;
}

export interface TournamentResultRecordedEvent {
  tournament: string;
  player: string;
  weight: bigint;
  elapsedSecs: bigint;
  moveCount: number;
  timestamp: number;
}

export interface PrizeClaimedEvent {
  tournament: string;
  player: string;
  amount: bigint;
  timestamp: number;
}
